---
title: "BIOSTAT M280 HW2"
author: "Chad Pickering"
date: "2/16/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Q1

### Exercise 7.3.4

*Question 1*  

*Explore the distribution of each of the `x`, `y`, and `z` variables in `diamonds`. What do you learn? Think about a diamond and how you might decide which dimension is the length, width, and depth.*  

We first plot the frequencies of measurements (in mm) of the three dimensions. We notice that the Z variable is concentrated between 2.5-5mm, whereas X and Y are very highly correlated, following almost exactly the same frequencies for the entire domain. The conclusion is that *Z is very likely the depth (height) of the diamond*, since from a birds' eye view, the length and width are essentially the same.   

```{r, message=FALSE}
library(tidyverse)
ggplot(diamonds) + 
  geom_freqpoly(mapping = aes(x = x, color = "X"), binwidth = 0.05) +
  geom_freqpoly(mapping = aes(x = y, color = "Y"), binwidth = 0.05) +
  geom_freqpoly(mapping = aes(x = z, color = "Z"), binwidth = 0.05) +
  coord_cartesian(xlim = c(2.5, 9)) +
  labs(title = "Frequency of Measurements of Dimensions of Diamonds",
       subtitle = "Bin width = 0.05mm",
       x = "Measurement (mm)",
       y = "Frequency") +
  theme(legend.position = "bottom") +
  scale_colour_discrete(name = "Dimension")
```

The following is a scatterplot of the length and width variables colored by the depth of the diamond. Despite a handful of points that do not follow the general approximate length=width trend (a few extreme outliers are obscured from view - these seem like data entry errors). As length/width increase, so does depth, quite linearly at that. We assert that X is length and Y is width - these terms can be interchanged, but the meaning will not be lost. An interesting observation is that out of 53940 diamonds, only 17 of them have an equal length and width.  

```{r}
z_common <- diamonds %>% 
  filter(z > 2 & z < 6) 
ggplot(z_common) +
  geom_point(mapping = aes(x = x, y = y, color = z)) + 
  coord_cartesian(xlim = c(3, 11), ylim = c(3, 11)) +
  labs(title = "Length vs. Width of Diamonds by Depth (mm)",
       x = "X (length) (mm)",
       y = "Y (width) (mm)") +
  theme(legend.position = "right") +
  scale_colour_continuous(name = "Z (depth)")
```

*Question 2*  

*Explore the distribution of `price`. Do you discover anything unusual or surprising? (Hint: Carefully think about the `binwidth` and make sure you try a wide range of values.)*  

We first plot the frequencies of prices with a bin width of 500 dollars. We see a positively skew, with a peak at about 1000 dollars and a gradual decreasing trend with few diamonds costing over 15000 dollars. Let's decrease the bin width to 100 dollars.  

```{r}
ggplot(diamonds) + 
  geom_freqpoly(mapping = aes(x = price), binwidth = 500) +
  labs(title = "Frequency of Prices of Diamonds",
       subtitle = "Bin width = $500",
       x = "Cost (dollars)",
       y = "Frequency")
```

With the bin width decreased to 100 dollars, we see that around 1500 dollars, very few observations are recorded, which is very strange, given that 1000 observations are recorded per bin in regions adjacent to it. Let's isolate that part of the dataset and confirm.  

```{r}
price_zoom_out <- ggplot(diamonds) + 
  geom_freqpoly(mapping = aes(x = price), binwidth = 100) +
  labs(title = "Frequency of Prices of Diamonds",
       subtitle = "Bin width = $100",
       x = "Cost (dollars)",
       y = "Frequency")

price_zoom_in <- ggplot(diamonds) + 
  geom_freqpoly(mapping = aes(x = price), binwidth = 100) +
  labs(title = "Frequency of Prices of Diamonds",
       subtitle = "Bin width = $100; zoomed to region of interest",
       x = "Cost (dollars)",
       y = "Frequency") +
  coord_cartesian(xlim = c(500, 2500))
  
require(gridExtra)
grid.arrange(price_zoom_out, price_zoom_in, nrow=2)
```

Indeed, below we see that there are no diamonds with prices between 1455 and 1545 dollars in the dataset. No other anomalies are spotted.  

```{r}
price_order <- diamonds %>%
  filter(price > 1453 & price < 1547) %>%
  select(price, x, y, z) %>%
  arrange(price)
head(price_order, 10)
```

*Question 3*  

*How many diamonds are 0.99 carat? How many are 1 carat? What do you think is the cause of the difference?*  

There are 23 0.99 carat diamonds in the dataset, while there are 1558 1 carat diamonds. Let's investigate possible causes of the difference - are 0.99 carat diamonds more rare? What are their properties in comparison with the 1 carat diamonds?  

```{r}
diamonds$xy_dist <- abs(diamonds$x - diamonds$y)

carat_99 <- diamonds %>%
  filter(carat == 0.99) %>%
  select(carat, price, x, y, z, xy_dist, cut, clarity)
carat_1 <- diamonds %>%
  filter(carat == 1) %>%
  select(carat, price, x, y, z, xy_dist, cut, clarity)
ct_lbl <- c("0.99 carat", "1 carat")
ct_count <- c(nrow(carat_99), nrow(carat_1))
ct_df <- as.data.frame(rbind(ct_lbl, ct_count))
row.names(ct_df) <- NULL
colnames(ct_df) <- NULL
ct_df
```

Both types of diamond have insignificant differences in their mean dimensions, but their mean price is significantly different (this is most likely a result, at least in part, of the fact that the carat value is not ideal, etc.). Let's see if the distributions of categorical cut and clarity measures are different between them.  

```{r}
t.test(carat_99$price, carat_1$price)
```

Here, we see that the proportion of 0.99 carat diamonds with lower quality cuts are higher than that of 1 carat diamonds.  *FINISH PLOT*

```{r}
carat_int <- rbind(carat_99, carat_1)
carat_int$carat <- as.factor(carat_int$carat)
ggplot(data = carat_int) + 
  geom_bar(mapping = aes(x = cut, fill = carat, y = ..prop..), position = "dodge")
```

*Question 4*  

Compare and contrast `coord_cartesian()` vs `xlim()` or `ylim()` when zooming in on a histogram. What happens if you leave binwidth unset? What happens if you try and zoom so only half a bar shows?  

The following uses `coord_cartesian()` with an unset binwidth. In this case, it defaults to 30. Let's zoom in to the domain [0, 5000] so that only half of a bar should show.  

```{r}
ggplot(diamonds) + 
  geom_histogram(mapping = aes(x = price)) +
  labs(title = "Frequency of Prices of Diamonds",
       subtitle = "Bin width = $500",
       x = "Cost (dollars)",
       y = "Frequency")
```

This seems to preserve the bin width to the default of the original histogram.  

```{r}
ggplot(diamonds) + 
  geom_histogram(mapping = aes(x = price)) +
  labs(title = "Frequency of Prices of Diamonds",
       subtitle = "Bin width = $500",
       x = "Cost (dollars)",
       y = "Frequency") +
  coord_cartesian(xlim = c(0, 5000))
```

Using xlim() or ylim() removes rows if those observations are not shown in the plot. Additionally, the bin widths are rescaled to fit the new domain, with a default of 30.  

```{r}
ggplot(diamonds) + 
  geom_histogram(mapping = aes(x = price)) +
  labs(title = "Frequency of Prices of Diamonds",
       subtitle = "Bin width = $500",
       x = "Cost (dollars)",
       y = "Frequency") +
  xlim(0, 5000)
```

### Exercise 7.4.1

*Question 1*  

*What happens to missing values in a histogram? What happens to missing values in a bar chart? Why is there a difference?*  

In a histogram, the rows where missing values are present are removed, and these values are not displayed.  

```{r}
diamonds2 <- diamonds %>% 
  mutate(z = ifelse(z < 2 | z > 6, NA, z))

all_z <- ggplot(diamonds) + 
  geom_histogram(mapping = aes(x = z), binwidth = 0.05) +
  labs(title = "Frequency of Heights of Diamonds",
       subtitle = "All z values included",
       x = "Depth (mm)",
       y = "Frequency") +
  coord_cartesian(xlim = c(5.5, 6.5), ylim = c(0, 50))

most_z <- ggplot(diamonds2) + 
  geom_histogram(mapping = aes(x = z), binwidth = 0.05) +
  labs(title = "Frequency of Heights of Diamonds",
       subtitle = "All z>6 excluded",
       x = "Depth (mm)",
       y = "Frequency") +
  coord_cartesian(xlim = c(5.5, 6.5), ylim = c(0, 50))

grid.arrange(all_z, most_z, nrow=2)
```

Here, we set all fair cut diamonds' "Cut" variable to NA and create bar plots. The category "Fair" is removed and replaced with the "NA" category, which has an identical count as "Fair" from the original dataset. This is because "NA" becomes an ordinal category, a factor; it is not removed from consideration in a bar plot.  

```{r}
diamonds3 <- diamonds %>% 
  mutate(cut = replace(cut, cut=="Fair", NA))

all_cut <- ggplot(data = diamonds) + 
  geom_bar(mapping = aes(x = cut)) +
  labs(title = "Counts of Types of Diamond Cut",
       subtitle = "All cuts included",
       x = "Cut",
       y = "Count")

most_cut <- ggplot(data = diamonds3) + 
  geom_bar(mapping = aes(x = cut)) +
  labs(title = "Counts of Types of Diamond Cut",
       subtitle = "Fair cuts excluded",
       x = "Cut",
       y = "Count")

grid.arrange(all_cut, most_cut, nrow=2)
```

*Question 2*  

What does `na.rm = TRUE` do in `mean()` and `sum()`?  

Let's take the first 500 values of `x` in the length column and set all of the values over 6 to NA. Taking the mean or sum of the vector that includes the NAs gives NA because when adding NA into the sum, the result is not a numerical value. However, the denominator of the mean still counts all values, including the NAs. The function `na.rm()` strips all NA values from consideration, reducing the count of values in the mean as well.  

```{r}
x_vec <- diamonds$x[1:500]
x_vec_na <- x_vec
x_vec_na[x_vec_na > 6] <- NA
c(mean(x_vec_na), sum(x_vec_na)) # na.rm = FALSE
c(mean(x_vec_na, na.rm = TRUE), sum(x_vec_na, na.rm = TRUE))
```

### Exercise 7.5.1.1

*Use what youâ€™ve learned to improve the visualisation of the departure times of cancelled vs. non-cancelled flights.*  

In the improved visualization, we use superimposed density curves to analyze the approximate relative frequency of departing (non-cancelled) flights compared to cancelled flights. We see given a flight is cancelled, it is likely that its departure time was between about 2-8pm rather than earlier in the day. We can see that the proportion of non-cancelled flights is bimodal, peaking in the early-mid morning, and again later in the day, while there are a low relative proportion of cancelled flights early in the day and a high relative proportion later in the day (compared to non-cancelled flights).  

```{r}
library(nycflights13)

# # Density - frequency-based
# flights %>%
#   mutate(
#     cancelled = is.na(dep_time),
#     sched_hour = sched_dep_time %/% 100,
#     sched_min = sched_dep_time %% 100,
#     sched_dep_time = sched_hour + sched_min / 60
#   ) %>% 
#   ggplot(mapping = aes(x = sched_dep_time, y = ..density..)) + 
#     geom_freqpoly(mapping = aes(colour = cancelled), binwidth = 1/2)

# Density - smoothed (preferred)
flights %>%
  mutate(
    cancelled = is.na(dep_time),
    sched_hour = sched_dep_time %/% 100,
    sched_min = sched_dep_time %% 100,
    sched_dep_time = sched_hour + sched_min / 60
  ) %>% 
  ggplot(mapping = aes(x=sched_dep_time, fill = cancelled)) +
    geom_density(stat = "density", alpha = 0.4, position = "identity") +
    labs(title = "Densities of Cancelled and Non-cancelled Flights",
       subtitle = "Proportions relative to groups",
       x = "Hour",
       y = "Density")
```

### Exercise 7.5.2.1

*How could you rescale the count dataset above to more clearly show the distribution of cut within colour, or colour within cut?*  

```{r}
diamonds %>% 
  count(color, cut) %>%  
  ggplot(mapping = aes(x = color, y = cut)) +
    geom_tile(mapping = aes(fill = n))
```

### Exercise 7.5.3.1

## Q3

Redo HW1 Q2 using functions in tidyverse:  

### Question 1.  

**How many persons are in the data set (statisticians call this `n`)?**  

Assuming each row is a unique person, we find **959** people (the command `uniq` is used anyway, just in case).  

**How many SNPs are in the data set (statisticians call this `p`)?**  

Assuming each row is a unique SNP, we find **8348674** SNPs (the command `uniq` is used anyway, just in case).

### Question 2.  

**Which chromosomes does this data set contain?**  



**How many SNPs are in each chromosome?** 



### Question 3.  

**MAP4 (microtubule-associated protein 4) is a gene on chromosome 3 spanning positions 47,892,180 bp -- 48,130,769 bp. How many SNPs are located within MAP4 gene?**  



**894** SNPs are located within the MAP4 gene.  

### Question 4.

**Statistical geneticists often have to reformat a data set to feed into various analysis programs. For example, to use the Mendel software we have to reformat the data set to be read by Mendel. Mendel's SNP definition file is similar to the plink `bim` file but has format `SNP ID`, `Chromosome`, `Base Pair Position` with each field separated by a comma. Write a Linux shell command to convert `merge-geno.bim` to Mendel SNP definition file.**   
      

    
**Mendel's pedigree file is similar to the plink `fam` file but has format `Family ID`, `Person ID`, `Father ID`, `Mother ID`, `Sex` coded as M or F, `Twin Status` with each field separated by a comma. Write a Linux shell command to convert `merge-geno.fam` to Mendel pedigree file. Since twin status is not available in plink format, we put nothing for that field. Also Mendel limits Person ID to have length less than or equal to 8 characters, so we have to strip the string `T2DG` from the IDs.**  






